Fix colors of plugin icons.

jasmussen opened the pull request:
This PR does two things:

1. It forcefully colorizes any SVG icon used in the toolbar or More menu. This allows us to be sure the contrast is present, and that we can show a toggled and untoggled color.

2. It adds margin to plugin icons, this fixes #7464.

Yes. I'm using `!important` here. I know you're not supposed to do that. But aside from being the only way we can ensure the colors, I feel like this is actually a valid usecase for the modifier. 

Screenshots:

![button color](https://user-images.githubusercontent.com/1204802/41772945-c4e7821a-761a-11e8-9e92-41c7def21329.gif)

<img width="253" alt="screen shot 2018-06-22 at 12 48 51" src="https://user-images.githubusercontent.com/1204802/41772947-c674dcb8-761a-11e8-9668-fc74e8b1c87d.png">

Commit:
Fix colors of plugin icons.

This PR does two things:

1. It forcefully colorizes any SVG icon used in the toolbar or More menu. This allows us to be sure the contrast is present, and that we can show a toggled and untoggled color.

2. It adds margin to plugin icons, this fixes #7464.

youknowriad commented on components/menu-item/style.scss:
```diff
@@ -27,6 +28,14 @@
+	g,
+	path {
+		stroke: $dark-gray-500 !important;
+		fill: $dark-gray-500 !important;
```
Plugins can still override using `!important`. I'm not certain we should/can enforce that personally.

jasmussen replied:
The benefit of us doing this as strongly as we can, is that the plugin developer doesn't have to do anything on order to get the icon to work on a dark background.

While I don't feel super strongly here, I think we should _start_ out opinionated, and then we can consider pulling back if a valid use case presents itself.

aduth commented on components/menu-item/style.scss:
```diff
@@ -27,6 +28,14 @@
+	// Forcefully colorize plugin icons to ensure contrast and cohesion.
+	svg,
+	g,
+	path {
```
The choice of elements to target here seems ad-hoc and incomplete, and I worry it could result in some abominations of icons where only parts are recolored (and not other parts such as `circle`, `ellipsis`, `line`, etc).

Example: https://codepen.io/aduth/pen/qKyxXJ

Would a wildcard `*` descendent selector be better?

jasmussen replied:
I dislike using !important as much as the next one. But most SVGs use inline styles for strokes and colors, and if we don't use !Iimportant, it means that the icon will have whatever color the plugin author provides, _including_ when the icon is toggled, where it should be white on the gray toggle background. In other words, if an SVG has `fill: green;` as an inline style, it will be green in every single context, hover, toggled, active etc.

jasmussen replied:
As an example:

![screen shot 2018-06-25 at 15 47 37](https://user-images.githubusercontent.com/1204802/41853877-2303ddf0-788f-11e8-8f9e-4650a6fa783e.png)

That's when I remove the !important from the is-toggled CSS class.

youknowriad replied:
@jasmussen I thought I had styles for that in the plugin itself (white path when toggled :P )

youknowriad replied:
As a "plugin developer" I think I agree with Gutenberg forcing a white path when toggled but I also think it shouldn't force a gray color when untoggled. Plugins should only care about providing the right colors for the default state (untoggled).

jasmussen replied:
I'd be okay with forcing colors when toggled.

aduth replied:
To clarify, I have no issue with forcing the color, just that there are more colorable SVG elements than just `svg`, `g`, and `path`.

mtias replied:
@jasmussen sorry I missed the follow ups here. I agree with forcing on toggle state, but I think e should leave the colors the plugin chose for the normal state.

jasmussen committed:
Try and address feedback.

jasmussen commented:
Okay, took a new stab at addressing the feedback. Here's how it looks now:

![icons](https://user-images.githubusercontent.com/1204802/41892217-9fc54198-7917-11e8-8705-04eab680c813.gif)

I'm targetting `svg, svg *` as the selectors for coloring. What happens now, as you can see starting https://github.com/WordPress/gutenberg/pull/7470/commits/66de82bac9e5b2bc1ce013e7ffe8a098288220a4#diff-cbe9c99c74fcdc63180d60c4535ed4eeR8 is this:

- We colorize SVG icons provided by the user. Not forcefully, but we do this so if a user downloads a material icon from https://material.io/icons/ and doesn't do anything to it, at least it's gray — those icons come without inline styles. But we don't do it forcefully, so if like in this example one icon is green and the other is purple, the plugin author can do that.
- We forcefully colorize the toggled state. This ensures the icon can be seen when toggled. 
- We forcefully colorize the hover state, this ensure that there * is * a hover state — without this, the base style would still shine through. In addition to this, it is necessary in order to be able to hover a toggled button.

Thoughts?

tofumatt reviewed (commented):
Wondering about Firefox SVG stuff… maybe I'm forgetting what works and what doesn't, but I recall having issues colouring SVGs back in the day...

tofumatt commented on components/menu-item/style.scss:
```diff
@@ -27,6 +28,19 @@
+	svg,
+	svg * {
+		stroke: $dark-gray-500;
+		fill: $dark-gray-500;
```
Update: hover works in Firefox but the **Dropit** plugin shows colours unless I force `!important` on more SVG selectors:

![2018-06-26 17 39 44](https://user-images.githubusercontent.com/90871/41926689-f30ef204-7967-11e8-8a82-0cf2ba0fb845.gif)

jasmussen replied:
Per discussion in https://github.com/WordPress/gutenberg/pull/7470#discussion_r197419174, this is intended behavior. 

That is: if an SVG does not provide inline styles, our gray color will apply for the pinned non-toggled non-hovered icon.

Regardless of inline styles, the colors will be enforced on hover and when toggled.

tofumatt commented on components/menu-item/style.scss:
```diff
@@ -7,8 +7,9 @@
-	.components-menu-items__item-icon {
-		margin-right: 5px;
+	.components-menu-items__item-icon,
+	> span > svg { // Also target plugin icons that can have arbitrary classes.
```
This is a weird spot for a comment, can it go above the selector instead?

It doesn't need the "also" either (the comma implies that).

`// Target plugin icons that can have arbitrary classes by using an aggressive selector.` should do the trick!

tofumatt commented on components/menu-item/style.scss:
```diff
@@ -27,6 +28,19 @@
 		@include menu-style__focus;
 	}
+
+	// Colorize plugin icons to ensure contrast and cohesion, but allow plugin devs to override.
```
Let's say "developers" 😄

tofumatt commented on edit-post/components/header/pinned-plugins/style.scss:
```diff
@@ -4,4 +4,24 @@
 		margin-left: 4px;
 	}
+
+	// Colorize plugin icons to ensure contrast and cohesion, but allow plugin devs to override.
```
This is repeated, which seems strange. But if it's staying we should at least say "developers".

tofumatt commented on edit-post/components/header/pinned-plugins/style.scss:
```diff
@@ -4,4 +4,24 @@
+		fill: $dark-gray-500;
+	}
+	
+	// Forcefully colorize hover & toggled plugin icon states to ensure legibility and consistency.
```
I guess the ampersand is fine… but why not just write "and" 😄

tofumatt reviewed (approved):
Looks good to me, just some minor nits.

But it works in Firefox for me. Nice idea!

tofumatt commented on components/menu-item/style.scss:
```diff
@@ -7,8 +7,9 @@
-		margin-right: 5px;
+	.components-menu-items__item-icon,
+	> span > svg { // Also target plugin icons that can have arbitrary classes.
+		margin-right: 4px;
```
Oh, this is covered by our weird RTL transform thing, right? I forget how it works. I think this is okay, but I forget.

jasmussen replied:
CC: @yoavf it should, right?

yoavf replied:
Yes, RTLCSS will take care of that in RTL mode.

jasmussen replied:
Thank you!

Commit:
Address feedback.

jasmussen commented:
Love your feedback! All the typos have been fixed, and I'll merge when green, noting that if there are disagreements on the forcefullness, we can always revisit that.
